platform/ios/src/MGLMapView.mm

```diff
@@ -5375,33 +5374,50 @@ void onCameraDidChange(mbgl::MapObserver::CameraChangeMode mode) override {
     }
 
     void onWillStartLoadingMap() override {
-        [nativeView willStartLoadingMap];
+        [nativeView mapViewWillStartLoadingMap];
     }
 
     void onDidFinishLoadingMap() override {
-        [nativeView didFinishLoadingMap];
+        [nativeView mapViewDidFinishLoadingMap];
     }
 
-    void onDidFailLoadingMap() override {
-        [nativeView didFailLoadingMap];
+    void onDidFailLoadingMap(mbgl::MapObserver::ErrorType type, const std::string& message) override {
+        NSError *error;
+        NSString *errorDescription = @(message.c_str());
+        MGLErrorCode errorCode;
+        switch (type) {
+        case mbgl::MapObserver::ErrorType::ParseStyle:
+            errorCode = MGLErrorCodeParseStyleFailed;
+            break;
+        case mbgl::MapObserver::ErrorType::NotFound:
+            errorCode = MGLErrorCodeNotFound;
+            break;
+        case mbgl::MapObserver::ErrorType::LoadingStyle:
+            errorCode = MGLErrorCodeLoadStyleFailed;
+            break;
+        }
+        error = [NSError errorWithDomain:MGLErrorDomain code:errorCode userInfo:errorDescription ? @{
+            NSLocalizedDescriptionKey: errorDescription,
```

Comment:
> could we add more subclasses for these error types?

I don't have a strong opinion on this, though for this approach to work we'd need to deduct the exception type via `catch` - see the exception juggling in https://github.com/mapbox/mapbox-gl-native/blob/master/include/mbgl/util/string.hpp#L44-L51 - for all platform-specific codes that needed to handle this.